File: 算法/1807. 替换字符串中的括号内容/Evaluate.go
package _807__替换字符串中的括号内容

import "strings"

func evaluate(s string, knowledge [][]string) string {
	// 构建散列表
	memo := make(map[string]string)
	for _, v := range knowledge {
		memo[v[0]] = v[1]
	}
	// 循环替换
	result := &strings.Builder{}
	start := 0
	flag := false
	// 循环字符串
	for index, value := range s {
		if value == '(' {
			// 如果是'('则记录下标
			start = index
			flag = true
		} else if value == ')' {
			// 结束,把从start开始到现在index的字符串截取下来
			// 左闭右开
			info := s[start+1 : index]
			if v, ok := memo[info]; ok {
				result.WriteString(v)
			} else {
				result.WriteString("?")
			}
			flag = false
		} else if !flag {
			result.WriteRune(value)
		}
	}
	return result.String()
}

File: 算法/1807. 替换字符串中的括号内容/Evaluate.java
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class Evaluate {

    class Solution {
        public String evaluate(String s, List<List<String>> knowledge) {
            // 构建散列表
            Map<String, String> memo = knowledge.stream().collect(Collectors.toMap(e -> e.get(0), e -> e.get(1)));
            boolean flag = false;
            int start = 0;
            StringBuilder result = new StringBuilder();
            for (int i = 0; i < s.length(); i++) {
                char c = s.charAt(i);
                if (c == '(') {
                    start = i;
                    flag = true;
                } else if (c == ')') {
                    // 获取括号中的内容 左闭右开
                    String value = s.substring(start + 1, i);
                    if (memo.containsKey(value)) {
                        result.append(memo.get(value));
                    } else {
                        result.append('?');
                    }
                    flag = false;
                } else if (!flag) {
                    result.append(c);
                }
            }
            return result.toString();
        }
    }

}
